<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Project 1: Python Coding Exercises

_Authors: Joseph Nelson (DC) _

---

The following code challenges are drawn from common exercises used in technical interviews.

Please note that there may be several ways to approach each challenge. If you get stuck, try mapping out your approach in pseudocode first. Finally, while solutions to problems like these may be found online, remember that if you copy/paste code that you can't explain, you'll be missing out on the point of the project. The only way to truly learn a new skill is through practice, trial, and error - we can only help you improve by understanding where you are having trouble.

### Challenge 1: Largest Palindrome
A palindromic number reads the same both ways. For example, 1234321 is a palindrome. The largest palindrome made from the product of two two-digit numbers is 9009 = 91 × 99. Find the largest palindrome made from the product of two three-digit numbers. Afterward, write a brief explanation walking through your code's logic in markdown.

In [1]:
n = 0
for a in range(999,100, -1):
    for b in range(a, 100, -1):
        x = a*b
        if x > n:
            s = str(a*b)
            if s == s[::-1]:
                n = a*b
print(n)

906609


In [2]:
n

906609

To find the largest pallindrom for the product of two three-digit number, first, assign n as variable and set the value as zero. To find largest pallindrom, using for loop with the range from 999. Basically a is iterating from 999 to 100, decreasing by 1 on each loop. b is iterating from a (from upper loop) until 100. This loop will be running for each value of a from upper loop. converting the product from int to str so that we can reverse the number. Below is possible for string not for int, so it is converted to string. s the reverse string of check. We check here if reversed string and non-reversed string are same (pallindrome if same). Thus, the output will be 906609.




### Challenge 2: Summation of Primes
The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17. Find the sum of all the primes below 2,000. Afterward, write a brief explanation walking through your code's logic in markdown.

In [5]:
sum_ans=17
for i in range(11,2000):
    for j in range(2,int(i**0.5+1)):
        if i%j==0:
            break
    else:
        sum_ans+=i
print(sum_ans)

277050


To find the sum of all primes below 2,000 , name the sum_ans as the variable and set the sum of the prime number value to 17. using for loop from the range value from 11 to 2000 and j in range from value 2 to less than i**0.5 and add value 1. So to check if the value of i is a prime number, we will use modulus operator for i divisable with value of j, if is equal to zero, then it will break the code. else, if the value have remainder, it will print the value of sum_ans that equal to the range of prime number that has modulus, and sums all the prime number value. Thus, the output is 277050. 



### Challenge 3: Multiples of 3 and 5
If we list all of the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6, and 9. The sum of these multiples is 23. Find the sum of all the multiples of 3 and 5 below 1,000. Afterward, write a brief explanation walking through your code's logic in markdown.

In [4]:
def f(x): return x%3 == 0 or x%5 == 0
print(sum(filter(f, range(1, 1000))))

233168


Using the formula 1 + 2 + 3 + ... + n = n*(n+1)/2 instead of a code loop.

Sum of all numbers <1000 divisible by 3:
3 + 6 + 9 + ... + 999 = 3 * (1 + 2 + 3 + ... + 333) = 3 * 333*334/2 = 166833

Sum of all numbers <1000 divisible by 5:
5 + 10 + 15 + ... + 995 = 5 * (1 + 2 + 3 + ... + 199) = 5 * 199*200/2 = 99500

Adding these up, those numbers divisible by 3*5 = 15 are counted double. Therefore we need to subtract these once:

Sum of all numbers <1000 divisible by 15:
15 + 30 + 45 + ... + 990 = 15 * (1 + 2 + 3 + ... + 66) = 15 * 66*67/2 = 33165

Thus we get the result 166833 + 99500 - 33165 = 233168.

### Challenge 4: String Compressor
Implement a method to perform basic string compression using the counts of repeated characters. (This is called run-length encoding.) For example, the string "aabcccccaaa" would become a2b1c5a3. If the “compressed” string would not become smaller than the original string, your method should return the original string. You can assume the string has only uppercase and lowercase letters (a–z). Specify whether your solution is case sensitive or case insensitive and what you would need to change to make it the other. Afterward, write a brief explanation walking through your code's logic in markdown.

In [13]:
from collections import Counter


def compress(string):
    temp = Counter()
    result = " "
    for x in string:
        temp[x] += 1

    for key, value in temp.items():
        result += str(key) + str(value)
    return result


if __name__== '__main__':
    s = input("Enter the string:")
    print(compress(s))

Enter the string:syakirah
 s1y1a2k1i1r1h1


To get the string compress, import counter library and assign function call compress in string parameter. Assign temp variable and make counter as value and assign result and empty value to run as empty string. Using for loop, check the value of x in string and add 1 everytime there is same string. to combine the value and string, use for loop again to run the string and value of the string and return the result. To the test the code, use wrap it into a function and call it from under an if __name__ == '__main__' guard and it will compress the string and print the result. 

### *BONUS* Challenge: FizzBuzz
Write a program that prints all of the numbers from 1 to 100. For multiples of 3, instead of the number, print "Fizz;" for multiples of 5, print "Buzz." For numbers that are multiples of both 3 and 5, print "FizzBuzz." Afterward, write a brief explanation walking through your code's logic in markdown.

In [4]:

# Loop for 100 times i.e. range
for fizzbuzz in range(1,100):

# Number divisible by 15,(divisible
# by both 3 & 5), print 'FizzBuzz'
# in place of the number
    if fizzbuzz % 15 == 0:
        print("FizzBuzz")
        continue
    
# Number divisible by 3, print 'Fizz'
# in place of the number
    elif fizzbuzz % 3 == 0:
        print("Fizz")
        continue

# Number divisible by 5,
# print 'Buzz' in
# place of the number
    elif fizzbuzz % 5 == 0:
        print("Buzz")
        continue

# Print numbers
print(fizzbuzz)


Fizz
Buzz
Fizz
Fizz
Buzz
Fizz
FizzBuzz
Fizz
Buzz
Fizz
Fizz
Buzz
Fizz
FizzBuzz
Fizz
Buzz
Fizz
Fizz
Buzz
Fizz
FizzBuzz
Fizz
Buzz
Fizz
Fizz
Buzz
Fizz
FizzBuzz
Fizz
Buzz
Fizz
Fizz
Buzz
Fizz
FizzBuzz
Fizz
Buzz
Fizz
Fizz
Buzz
Fizz
FizzBuzz
Fizz
Buzz
Fizz
Fizz
99


Firstly, we need to use for loop to make it loop for 100 times. Next, to get the result of fizzbuzz, for both number 3 and 5, that is divisable by 15 equal to zero it will print FizzBuzz, if the result is false, it will turn to the condition 2, for the number 3 that can divisable by 3 print Fizz. If the result for condition 2 is false, it will run to the next condition which if is the number 15 can be divisible by 5 and equal to zero, it will print Buzz. Otherwise, if all condition is false until end of the loop, it will print the last number which is 99. 